pufferlib/ocean/torch.py

```diff
@@ -172,3 +172,118 @@ def decode_actions(self, flat_hidden, lookup, concat=None):
             #print('argmax samples: ', argmax_samples)
 
             return action, value
+
+
+class TrashPickup(nn.Module):
+    def __init__(self, env):
+        super().__init__()
+        self.num_trash = env.num_trash
+        self.num_bins = env.num_bins
+        self.num_agents_per_env = env.num_agents_per_env
+
+        self.trash_net = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(3 * env.num_trash, 32)),  # (presence, x, y) for each trash
+            nn.ReLU(),
+            pufferlib.pytorch.layer_init(nn.Linear(32, 16)),
+            nn.ReLU(),
+        )
+
+        self.bin_net = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(2 * env.num_bins, 8)),  # (x, y) for each bin
+            nn.ReLU(),
+            pufferlib.pytorch.layer_init(nn.Linear(8, 4)),
+            nn.ReLU(),
+        )
+
+        self.other_agent_net = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(3 * (env.num_agents_per_env - 1), 16)),  # (x, y, carrying) for each other agent
+            nn.ReLU(),
+            pufferlib.pytorch.layer_init(nn.Linear(16, 8)),
+            nn.ReLU(),
+        )
+
+        self.position_net = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(2, 8)),
+            nn.ReLU(),
+        )
+        
+        self.carrying_net = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(1, 2)),
+            nn.ReLU(),
+        )
+
+        self.proj = nn.Sequential( 
+            nn.Linear(16 + 4 + 8 + 8 + 2, 32),  # Combined features size
+            nn.ReLU(),
+            nn.Linear(32, 16),
+            nn.ReLU(),
+        )
+
+        self.actor = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(16, 8)),
+            nn.ReLU(),
+            pufferlib.pytorch.layer_init(nn.Linear(8, 4), std=0.01),  # 4 actions
+        )
+
+        self.critic = nn.Sequential(
+            pufferlib.pytorch.layer_init(nn.Linear(16, 8)),
+            nn.ReLU(),
+            pufferlib.pytorch.layer_init(nn.Linear(8, 1), std=1),  # Value prediction
+        )
+
+        self.is_continuous = isinstance(env.single_action_space, pufferlib.spaces.Box)
+
+    def forward(self, observations):
+        hidden, _ = self.encode_observations(observations)
+        actions, value = self.decode_actions(hidden)
+        return actions, value
+
+    def encode_observations(self, observations):
+        """
+        Encode observations for each agent.
+        """
+
+        obs_index = 0
+
+        # Agent's own position and carrying status
+        agent_position = observations[:, obs_index : obs_index + 2]
+        carrying_status = observations[:, obs_index + 2 : obs_index + 3]
+        obs_index += 3
+
+        # Other agents
+        other_agents = observations[:, obs_index : obs_index + 3 * (self.num_agents_per_env - 1)]
```

Comment:
Not sure what is going on here. There shouldn't be an obs for every agent seeing every other agent in most cases. Usually, you want a 2d crop of tiles, so I'm surprised to even see 1d data here

Reply:
Completed (to be pushed).